### PR TITLE
new recipe: libyang

### PR DESCRIPTION
**Pull request by tbsuht**
Specify library name and version:  **libyang/2.1.148**

[YANG](https://en.wikipedia.org/wiki/YANG) is a data modeling language and `libyang` implements the specified RFCs.
This is e.g. heavily used in the telecommunication/networking area to store configuration/state data of a device.

There are other projects for which I want to also contribute recipes which depend on `libyang`.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libyang/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:44]:
```suggestion
        tc = CMakeToolchain(self)
        tc.variables["ENABLE_TESTS"] = False
        tc.variables["ENABLE_VALGRIND_TESTS"] = False
        tc.variables["ENABLE_STATIC"] = not self.options.shared
```
Please, disable tests: https://github.com/CESNET/libyang/blob/v2.1.148/CMakeLists.txt#L235

File: recipes/libyang/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+from conan import ConanFile
+from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout, CMakeDeps
+from conan.tools.files import get
+
+required_conan_version = ">=1.53.0"
+
+
+class LibYangConan(ConanFile):
+    name = "libyang"
+    license = "BSD-3-Clause license"
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:10]:
```suggestion
    license = "BSD-3-Clause"
```

Use the short form listed in SPDX: https://spdx.org/licenses/BSD-3-Clause.html

File: recipes/libyang/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.set_property("cmake_find_mode", "both")
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:59]:
```suggestion
```
We do not enforce find mode, unless really needed.

File: recipes/libyang/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        self.cpp_info.set_property("cmake_file_name", "libyang")
+        self.cpp_info.libs = ["yang"]
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:61]:
```suggestion
        self.cpp_info.libs = ["yang"]
        if self.settings.os == "Windows":
            self.cpp_info.system_libs.extend(["ws2_32", "shlwapi"])
        elif self.cpp_info.os in ["Linux", "FreeBSD"]:
            self.cpp_info.system_libs.extend(["pthread", "m"])
```
Need to list required system libraries. 

https://github.com/CESNET/libyang/blob/v2.1.148/CMakeLists.txt#L395
https://github.com/CESNET/libyang/blob/v2.1.148/CMakeLists.txt#L387

File: recipes/libyang/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        tc.generate()
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.install()
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:56]:
```suggestion
        cmake.install()
        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
```

Need to remove generated files and use only from Conan generator pkg_config

**Review by uilianries — COMMENTED** [review]
@tbsuht Thank you for your PR! I didn't test your PR locally, but I looked into the upstream's CMakeLists.txt. Please, check my review.

**Comment by tbsuht** [comment]
@uilianries `Unknown folder 'share' in the package`. Some files in there are useful/required for dependent projects. Should I move them to `res`?

File: recipes/libyang/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+    def package(self):
+        copy(self, "LICENSE", self.source_folder,
+             os.path.join(self.package_folder, "licenses"))
+        cmake = CMake(self)
+        cmake.install()
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+
+    def package_info(self):
+        self.cpp_info.set_property("cmake_file_name", "libyang")
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:67]:
```suggestion
        self.cpp_info.set_property("cmake_file_name", "LibYANG")
```

https://github.com/CESNET/libyang/blob/master/FindLibYANG.cmake

File: recipes/libyang/all/test_package/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package C)
+
+find_package(libyang REQUIRED)
```
Comment (uilianries) [inline recipes/libyang/all/test_package/CMakeLists.txt:4]:
```suggestion
find_package(LibYANG REQUIRED)
```

File: recipes/libyang/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version],
+            destination=self.source_folder, strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["ENABLE_TESTS"] = False
+        tc.variables["ENABLE_VALGRIND_TESTS"] = False
+        tc.variables["ENABLE_STATIC"] = not self.options.shared
```
Comment (valgur) [inline recipes/libyang/all/conanfile.py:51]:
```suggestion
        tc.variables["ENABLE_STATIC"] = not self.options.shared
        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0077"] = "NEW"
```
Needed for correct functioning of `tc.variables` with `cmake_minimum_required(VERSION 2.8.12)`.

File: recipes/libyang/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+               "fPIC": [True, False]}
+    default_options = {
+        "shared": False,
+        "fPIC": True
+    }
+
+    tool_requires = "cmake/[>=3.22.0 <4]"
+
+    def validate(self):
+        # TODO support windows build
```
Comment (tbsuht) [inline recipes/libyang/all/conanfile.py:29]: I don't have a Windows machine available to get this running

File: recipes/libyang/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration(
+                f"{self.ref} on Windows is not yet supported.")
+
+    def requirements(self):
+        self.requires("pcre2/10.42", transitive_headers=True)
+        # TODO windows build
```
Comment (tbsuht) [inline recipes/libyang/all/conanfile.py:36]: see above

File: recipes/libyang/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        "fPIC": True
+    }
+
+    tool_requires = "cmake/[>=3.22.0 <4]"
+
+    def validate(self):
+        # TODO support windows build
+        if self.settings.os == "Windows":
+            raise ConanInvalidConfiguration(
+                f"{self.ref} on Windows is not yet supported.")
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:32]:
```suggestion
        # TODO For Windows support: https://github.com/CESNET/libyang?tab=readme-ov-file#windows-build-requirements
        # CMake Error at CMakeLists.txt:386 (find_package):
        #   By not providing "Findpthreads.cmake" in CMAKE_MODULE_PATH this project has
        if self.settings.os == "Windows":
            raise ConanInvalidConfiguration(f"{self.ref} Conan recipe is not prepared to work on Windows. Contributions are welcome.")
```

File: recipes/libyang/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    def requirements(self):
+        self.requires("pcre2/10.42", transitive_headers=True)
+        # TODO windows build
+        # if is_msvc(self):
+        #     self.requires("getopt-for-visual-studio/20200201")
+        #     self.requires("dirent/1.24")
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:39]:
```suggestion
```

I would not include for now.

File: recipes/libyang/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        cmake.build()
+
+    def package(self):
+        copy(self, "LICENSE", self.source_folder,
+             os.path.join(self.package_folder, "licenses"))
+        cmake = CMake(self)
+        cmake.install()
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+        # move /share to /res
+        copy(self, "*", os.path.join(self.package_folder, "share"),
```
Comment (uilianries) [inline recipes/libyang/all/conanfile.py:80]: What content is so important from share folder? If is only manpage and docs stuff, I would prefer removing it.

**Review by uilianries — COMMENTED** [review]
Almost good to go, only few comments.

File: recipes/libyang/all/conanfile.py (hunk 13 above)
Comment (tbsuht) [reply recipes/libyang/all/conanfile.py]: I have restricted it to the contained *.yang files

**Review by ErniGH — COMMENTED** [review]
Thanks for your pull request ❤️

**Review by ErniGH — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Review by RubenRBS — APPROVED** [review]

**Pull request merged**